### PR TITLE
[FPD] Add Autobrake to FMA

### PR DESCRIPTION
**Pull request by MisterChocker**
Fixes #[issue_no]

## Summary of Changes

- Added BRAKE MAX, MED and LO to FMAs
- Restructured FMA to support Brake FMAs
- Fixed bad condition for GA Soft (wrong flight phase)
- Added Decel to PFD

DECEL becomes available 4 secs (BRK LO) or 2 secs (BRK MED) after main landing gear touch down.
DECEL becomes active when available and deceleration noticeable.

## Additional context

Discord username (if different from GitHub): DerL30N#3751

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Review by Benjozork — APPROVED** [review]
LGTM

**Comment by valenzz2004** [comment]
Category: QA Tester
Name: Valenzz2004#7802
Date of testing: 20/12/2020
Version of the sim: 1.11.6.0
PR Tested: #2483 
Object of test: Test the added autobrake to FMA

Testing technique: to complete a full flight to see if functions were correctly implemented
Steps you've made to test the Pull Request:

1. Loaded in at a stand at EGLL
2. setup aircraft and imported a flight plan via simbrief (EGLL-EGCC)
3. taxied to the active runway (autobrake max selected and identified)
4. departed and reached CRZ
5.  at 6000 entered dest data armed spoilers and autobrake low (most appropriate for 23R at EGCC)
6. landed on the ILS 23R absolutely no conflicts with other functions of the A32NX

Flight notes (If required):

Issues (If Any):
Overall Rating: Good
Conclusions of the testing: worked as i expected and i also noticed autobrake max is not selectable in the air so should be one step closer to the real aircraft :)
Screenshots( if any):

Before departure
QA ground test identifier

Final
short final

On touchdown decel
decel on landing

**Comment by sidnov** [comment]
Category: QA Tester
Name: sidnov#8337
Date of testing: 20/12/2020
Version of the sim: 1.11.7.0
PR Tested: #2483
Object of test: Autobrake indication on FMA
Testing technique: Runway spawn and circuit flying
Steps you've made to test the Pull Request:
1. Spawned at VABB RWY27 for departure. BRK MAX present on the FMA as per TO CONFIG.
2. Took off, flew a circuit and back to land at RWY27.
3. Modes BRK MED and BRK LO available to arm on the approach
4. Landed with BRK LO mode

Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: Indications work as expected. As Valenzz2004#7802 pointed out, BRK MAX cannot be enabled mid-air, like the real aircraft. Looks good to go!


Screenshot 2020-12-20 152648
Screenshot 2020-12-20 152740

**Comment by BritishAvgeek** [comment]
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: 20/12/2020
Version of the sim: 1.11.7.0
PR Tested: #2483 
Object of test: Autobrake added to FMA

Testing technique: Circuits at EGKK - 3 touch and goes and 1 landing
Steps you've made to test the Pull Request:
1. Spawned Cold and Dark
2. Flew 3 circuits with touch and goes - each time arming the autobrake to Lo, Med, and Lo for each consecutive circuit to test each one
3. Fourth circuit the setting was changed numerous times to see if it would stick. Med -> Lo -> Med -> Off -> Lo -> Med -> Lo - Off -> Med then back to Lo for approach and a full landing was completed
4. No other issues detected

Flight notes (If required): N/A

Issues (If Any): None found during any tests. Didn't cause issues elsewhere in the mod from what I was looking at, and the autobrake settings displayed correctly each time - even on the fourth circuit when I was trying to break it
Overall Rating: Good
Conclusions of the testing: No issues found, all appears to be working well

**Comment by MisterChocker** [comment]
Thanks guys for the testing and especially actively trying to find bugs and even better to see you found none!
@BritishAvgeek short question to you, the autobrake has some "activation delay", IRL thats caused by a couple of factors to the activation system. In the sim I've simply set a delay of 4 seconds after main landing gear touchdown before showing decel and autobrake active (I took this time from different videos). So my question is, is it appropriate?

**Review by kberg93 — APPROVED** [review]
Category: QA Tester
Name: Kberg#0666
Date of testing: Current date 12/21/2020
Version of the sim: 1.11.7.0
PR Tested: #2483 
Object of test: [FPD] Add Autobrake to FMA

Testing technique: Just a Quick RTO

Steps you've made to test the Pull Request: 
1. Put in take off data on (hot start on rwy).
2. Speeded up to 120.
3. Rejected and braked max+ rev thrust.

Flight notes (If required): Just a quick RTO since no one else did try that :)

Issues (If Any): none

Overall Rating: Good

Conclusions of the testing: No issues screenshot of before and during RTO added under.

Screenshots:
FlightSimulator_iJbSQhIKEt
KfRWCD8bQe

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AirspeedIndicator.js [hunk 1]
```diff
@@ -1021,6 +1045,15 @@ class Jet_PFD_AirspeedIndicator extends HTMLElement {
             takeOffSpeedNotSet = true;
         }
 
+        if (this.decelText) {
+            if (_isDecel) {
+                this.decelText.setAttribute("visibility", "visible");
+                this.decelText.textContent = "DECEL";
+            } else {
+                this.decelText.setAttribute("visibility", "hidden");
+            }
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AirspeedIndicator.js:1054]: Can we diff with the previous value for performance reasons ?

**Comment by wpine215** [comment]
> Thanks guys for the testing and especially actively trying to find bugs and even better to see you found none!
> @BritishAvgeek short question to you, the autobrake has some "activation delay", IRL thats caused by a couple of factors to the activation system. In the sim I've simply set a delay of 4 seconds after main landing gear touchdown before showing decel and autobrake active (I took this time from different videos). So my question is, is it appropriate?

If I'm not mistaken, there is a delay of 4 seconds for autobrake low, and 2 seconds for autobrake medium. Not entirely sure about autobrake max, but considering it's used for rejected take-offs, I would assume it takes effect immediately.

**Comment by MisterChocker** [comment]
> If I'm not mistaken, there is a delay of 4 seconds for autobrake low, and 2 seconds for autobrake medium. Not entirely sure about autobrake max, but considering it's used for rejected take-offs, I would assume it takes effect immediately.

I've got the delay only for LO and MED due to the fact these are the only two activated on touchdown.
I can make the delay 2 and 4 sec depending on the config, do you have show me some references for me?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AirspeedIndicator.js (hunk 1 above)
Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AirspeedIndicator.js:1054]: yes we could, FYI the whole display doesn't do that atm

**Comment by wpine215** [comment]
> 
> 
> > If I'm not mistaken, there is a delay of 4 seconds for autobrake low, and 2 seconds for autobrake medium. Not entirely sure about autobrake max, but considering it's used for rejected take-offs, I would assume it takes effect immediately.
> 
> I've got the delay only for LO and MED due to the fact these are the only two activated on touchdown.
> I can make the delay 2 and 4 sec depending on the config, do you have show me some references for me?

Taken from FCOM:
image

**Comment by BritishAvgeek** [comment]
> Thanks guys for the testing and especially actively trying to find bugs and even better to see you found none!
> @BritishAvgeek short question to you, the autobrake has some "activation delay", IRL thats caused by a couple of factors to the activation system. In the sim I've simply set a delay of 4 seconds after main landing gear touchdown before showing decel and autobrake active (I took this time from different videos). So my question is, is it appropriate?

Sorry only just saw this! The delay seemed fine for me!

**Comment by beheh** [comment]
Is the DECEL message behavior identical to the green light on the autobreak buttons?

image

**Comment by MisterChocker** [comment]
> If not, could we make those two match?

No currently not, logic on autobrake button is wrong.

**Comment by Road-to-Hell9** [comment]
Category: QA Tester
Name: Road_to_Hell9 #1111
Date of testing: 21/12/2020
Version of the sim: Newes 21/12/2020
PR Tested: #2483
Object of test: [FPD] Add Autobrake to FMA

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1.Normal cold start
2.takeoff Checklist
3.Set Autobrakes to max
4.
5.

Flight notes (If required):
The display works

Issues (If Any): The Autobrakes just deactivate
Overall Rating: TO be improved
Conclusions of the testing: The autobrakes have to be prepared (No sure if it's your fault)

**Comment by MisterChocker** [comment]
@Road-to-Hell9 How got the autobrake deactivated? got the autobrake unintentionally engaged, also I am not disabling the autobrake in this pr, I am just displaying the autobrake status on the FMA/PFD

**Comment by Road-to-Hell9** [comment]
The autobrakes disable a few seconds after activation, during taxi and or holding. I said in my report "Not sure if it's your fault" I'm sorry if I offended you but I just wanted to report it, I don't see why you had to dislike my comment, I'm just trying my best.

**Comment by BritishAvgeek** [comment]
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: 20/12/2020
Version of the sim: 1.11.7.0
PR Tested: #2483
Object of test: Autobrake added to FMA - Test 2

Testing technique: Circuits at EGCC - 3 touch and goes and 1 landing - exact same technique as Test 1
Steps you've made to test the Pull Request:

1. Spawned Cold and Dark
2. flew 3 circuits with touch and goes - each time arming the autobrake to Lo, Med, and Lo for each consecutive circuit to test each one
3. Fourth circuit the setting was changed numerous times to see if it would stick. Med -> Lo -> Med -> Off -> Lo -> Med -> Lo - Off -> Med then back to Lo for approach and a full landing was completed
4. No other issues detected
5. Taxi back to the runway and perform a rejected Takeoff using Max autobrake

Flight notes (If required): N/A

Issues (If Any): Noted the above change to delay for 2seconds for Med being amended from the pair 4s with Lo. Both settings worked well with the delay at 2 seconds as expected for medium autobrake, and 4s for Low Autobrake. Max was instant during the RTO test and remained on until I took over braking control myself.
Overall Rating: Good
Conclusions of the testing: No issues found, all appears to be working well

**Pull request merged**